api/v1alpha2/conversion.go

```diff
@@ -247,15 +254,58 @@ func (dst *MachineDeploymentList) ConvertFrom(srcRaw conversion.Hub) error {
 }
 
 func Convert_v1alpha2_MachineSpec_To_v1alpha3_MachineSpec(in *MachineSpec, out *v1alpha3.MachineSpec, s apiconversion.Scope) error {
+	inGV, err := schema.ParseGroupVersion(in.Bootstrap.ConfigRef.APIVersion)
+	if err != nil {
+		return err
+	}
+
 	if err := autoConvert_v1alpha2_MachineSpec_To_v1alpha3_MachineSpec(in, out, s); err != nil {
 		return err
 	}
 
+	if inGV.Group == bootstrapv1a3.GroupVersion.Group && (out.Bootstrap.ConfigRef.Kind == kubeadmConfigKind || out.Bootstrap.ConfigRef.Kind == kubeadmConfigTemplateKind) {
+		out.Bootstrap.ConfigRef.APIVersion = bootstrapv1a3.GroupVersion.String()
+	}
+
 	// Discards unused ObjectMeta
 
 	return nil
 }
 
+func Convert_v1alpha2_MachineSetSpec_To_v1alpha3_MachineSetSpec(in *MachineSetSpec, out *v1alpha3.MachineSetSpec, s apiconversion.Scope) error {
+	inGV, err := schema.ParseGroupVersion(in.Template.Spec.Bootstrap.ConfigRef.APIVersion)
```

Comment:
This should be guarded against `in.Template.Spec.Bootstrap != nil`, move this at the bottom

Reply:
If `in.Template.Spec.Bootstrap == nil`, is it a normal case or an error condition? I earlier thought of doing this but somehow thought except for objects created in tests, the input `in` objects should always have non-nil `in.Template.Spec.Bootstrap`. But may be I am wrong.

Reply:
If the Bootstrap is nil that's a normal condition, and no further action is necessary